python/ray/data/impl/delegating_block_builder.py

```diff
@@ -26,13 +27,24 @@ def add(self, item: Any) -> None:
                     self._builder = ArrowBlockBuilder()
                 except (TypeError, pyarrow.lib.ArrowInvalid):
                     self._builder = SimpleBlockBuilder()
+            elif isinstance(item, np.ndarray):
+                self._builder = ArrowBlockBuilder()
             elif isinstance(item, PandasRow):
                 self._builder = PandasBlockBuilder()
             else:
                 self._builder = SimpleBlockBuilder()
         self._builder.add(item)
 
-    def add_block(self, block: Block) -> None:
+    def add_batch(self, batch: DataBatch):
+        """Add a user-facing data batch to the builder.
+
+        This data batch will be converted to an internal block and then added to the
+        underlying builder.
+        """
+        block = BlockAccessor.batch_to_block(batch)
+        return self.add_block(block)
+
+    def add_block(self, block: Block):
```

Comment:
I was going to defer this to de-risk the PR, since some of the datasources use `.add_block()`, so we could get rid of this if we also wanted to support read tasks being able to return batch data types (e.g. the NumPy datasource could return ndarrays and let the delegating block builder convert them to Arrow tables). Even though this seems nice, I opted to not do it since it is a scope creep.

Reply:
Sounds good. Next time could add a TODO in case we forget.